### PR TITLE
Improved BatchHolder interface to avoid memory lifetime error

### PR DESCRIPTION
**Pull request by bright-starry-sky**
I think I made a design mistake I shouldn't have. 

The definition of batch_:
`std::vector<std::pair<BatchLogType, std::pair<folly::StringPiece, folly::StringPiece>>> batch_;
`

I can't use it this way:
```
for (int a = 1; a < 3; a++ ) {
    auto key = folly::stringPrintf("key_%d", a);
     auto val = folly::stringPrintf("val_%d", a);
    batch_.emplace_back(BatchLogType::OP_BATCH_PUT,
                                          std::pair<folly::StringPiece, folly::StringPiece>(key, val));
}
```

Now the batch_'s elements as below, it is invalid data:
```
{
    {OP_BATCH_PUT, {key2, key2}},
    {OP_BATCH_PUT, {key2, key2}}
}
```

### CODE REVIEW

File: src/kvstore/test/NebulaStoreTest.cpp [hunk 1]
```diff
@@ -825,6 +826,100 @@ TEST(NebulaStoreTest, ThreeCopiesCheckpointTest) {
...
+    // put kv
+    {
+        std::vector<std::pair<std::string, std::string>> expected, result;
+        auto atomic = [&]() -> std::string {
+            std::unique_ptr<kvstore::BatchHolder> batchHolder =
+                    std::make_unique<kvstore::BatchHolder>();
+            for (auto i = 0; i < 20; i++) {
+                auto key = folly::stringPrintf("key_%d", i);
+                auto val = folly::stringPrintf("val_%d", i);
+                batchHolder->put(key, val);
```
Comment (dangleptr) [inline src/kvstore/test/NebulaStoreTest.cpp]: As your usage,  i suggest we use std::string as the argument type for batchHolder->put

File: src/kvstore/LogEncoder.cpp [hunk 2]
```diff
@@ -165,7 +165,7 @@ std::vector<folly::StringPiece> decodeMultiValues(folly::StringPiece encoded) {
 }
 
 std::string encodeBatchValue(const std::vector<std::pair<BatchLogType,
-                             std::pair<folly::StringPiece, folly::StringPiece>>>& batch) {
+                             std::pair<std::string, std::string>>>& batch) {
```
Comment (monadbobo) [inline src/kvstore/LogEncoder.cpp]: Make the parameter const or remove the reference and "move" the parameter in.

**Comment by bright-starry-sky** [comment]
Addressed @dangleptr's comment.

File: src/kvstore/LogEncoder.h [hunk 3]
```diff
@@ -64,33 +64,32 @@ class BatchHolder {
...
+                            std::make_pair(begin, end));
     }
 
     void clear() {
         batch_.clear();
     }
 
     std::vector<std::pair<BatchLogType,
-                     std::pair<folly::StringPiece, folly::StringPiece>>> getBatch() {
+                          std::pair<std::string, std::string>>> getBatch() {
```
Comment (dangleptr) [inline src/kvstore/LogEncoder.h]: return const reference？

File: src/kvstore/LogEncoder.h [hunk 4]
```diff
@@ -64,33 +64,32 @@ class BatchHolder {
     BatchHolder() = default;
     ~BatchHolder() = default;
 
-    void put(folly::StringPiece key, folly::StringPiece val) {
+    void put(std::string&& key, std::string&& val) {
         batch_.emplace_back(BatchLogType::OP_BATCH_PUT,
-                            std::pair<folly::StringPiece, folly::StringPiece>(key, val));
+                            std::make_pair(key, val));
```
Comment (dangleptr) [inline src/kvstore/LogEncoder.h]: std::make_pair(std::forward<std::string>(key),    std::forward<std::string>(val));

**Comment by bright-starry-sky** [comment]
Jenkins go

**Comment by critical27** [comment]
I guess it must be hard to find out 🤣, LGTM

**Pull request merged**